README.md
# DualBoot TWRP/OrangeFox for the Oneplus 6 series

A big thanks to Zackptg5. The master of this mod from the beginning. I just adjusted this now to work on the Oneplus 6 series phones.

Modified TWRP (Mauronofrio's build) and installer script for Oneplus 6 series that re-purposes userdata for true dual booting. You can still use this as a regular stock twrp zip - one stop shop for magisk, verity, and/or forced encryption modifications.

You can choose to install this Dualboot-mod with a blacked themed TWRP or with OrangeFox recovery.

## Changelog
* 05/14/2021 - 3.5.2-9 v6
  * Updated Dualboot Companion app to V2.9.1
  DualBoot companion App V2.9.1 Changelog
  * Optimized splash screen
  * Fixed reboot time on Custom Rom in widget menu 

* 04/09/2021 - 3.5.2-9 v5.3
  * Updated TWRP to 3.5.2-9 (black Themed)
  * Updated Dualboot Companion app to V2.8.9
  DualBoot companion App V2.8.9 Changelog
  * Added Splash screen at start
  * Optimized starting time
  * Added free space on CommonData
  * Donate button 
  * Update Russian translation
  * bug fix here and there
  * Fixed Shortcuts
  * Fixed colours in updated box
  * Updated Italian and Russian translation
  * new logo
  * new info section
  * fixed crash using shortcuts
  * shared app updated:
  * Data Restore feature (it will copy the shared data back to /data/data and dismount the bind mount)
  * Data Delete feature (it will delete the shared data and dismount the bind mount)
  * Info about shared data if they are in place in shared directory

* 03/18/2021 - 3.5.1-9 v5.2
  * Updated with TWRP 3.5.1-9 (black Themed)

* 03/15/2021 - 3.5.0-9 v5.1
  * Updated Dualboot Companion app to V2.8.5
  * Added Dualboot with OrangeFox Recovery R11.0
  * Added Dualboot Logo to OrangeFox Recovery
  * You can now choose which recovery you want to use with the Dualboot. TWRP or OrangeFox Recovery
  * Fixed bug in partition size details for 64GB Oneplus 6 devices
  * Fixed error when flashing a ROM in OrangeFox recovery
  * Fixed hanging/timeout when rebooting to System or Recovery on Android 11 (works only when using reboot function from inside the Dualboot Companion app)

* 03/05/2021 - 3.5.0-9 v5
  * Added Dualboot Companion app created by User Invernomut0
  * small bugfixes and corrections

* 02/07/2021 - 3.5.0-9 v4.4
  * Fixed bug which caused that the quickmode options were not working

* 01/17/2021 - 3.5.0-9 v4.3
  * Updated with Magisk 21.4

* 01/16/2021 - 3.5.0-9 v4.2
  * Updated with Magisk 21.3

* 01/08/2021 - 3.5.0-9 v4.1
  * Cleaned up stuff which is not needed anymore
  * Fixed bugs that were existing in the repartitioning part
  * Updated with official TWRP 3.5.0-9 and added black Theme to it
  * Added option to create a seperate partition called StockSD when on Stock Layout

* 12/29/2020 - 3.4.0-0 v4
  * Updated with Magisk 21.2
  * Removed Magisk Canary 20422 since Magisk 21.2 is now working with TWRP and Android 11

* 12/01/2020 - 3.4.0-0 v3.3
  * Added to flash Magisk Canary 20422 on Android 11 automatically
  * Added mount file to mount Android Data to Commondata (still Beta/Test)

* 11/14/2020 - 3.4.0-0 v3.2
  * Updated with Magisk 21.1
  * Prepared script for some future features

* 10/30/2020 - 3.4.0-0 v3.1
  * Removed unnessessary stuff which was left from the Oneplus 7 devices
  * Fixed Common data mount script for Android 11
  * Magisk 20.4
  * Dualboot-TWRP can be flashed from USB Storage as well now

* 10/20/2020 - 3.4.0-0 v3
  * Added Blackout TWRP Themed by acuicultor and Morphine1
  * Added support/detection for 64GB Oneplus 6 enchilada devices
  * Added that Magisk 21.0 does not get flashed on Android 11 ROMs (as that is not working now!)
  * removed automatic launch of Common data mount script on Android 11 as that caused strange behaviour when rebooting the phone!
  * Some more small adjustments in the flashing script
  * Adjusted sizes for the userdata partitions when using A/B/C Layout as follows: 
    64GB    16, 24 GB partition size for A/B/C Layout
	128GB   16, 32, 40, 45 GB partition size for A/B/C Layout
	256GB   16, 32, 64, 96 GB partition size for A/B/C Layout

* 10/17/2020 - 3.4.0-0 v2.3
  * Fixed mounting of Common data partition
  
* 10/15/2020 - 3.4.0-0 v2.2
  * Fixed better detection of Android 11 Installed ROMS
  * Build Dualboot Installer for Oneplus 6 enchilada

* 10/13/2020 - 3.4.0-0 v2.1
  * Fixed hanging when on Android 11

* 10/12/2020 - 3.4.0-0 v2
  * Fixed automatic Formatting and creating of the partitions
  * added quickmode options which can be provided directly to the ZIP file name
  * A lot of other bugfixes and improvments
  
* 10/06/2020 - 3.4.0-0 v1
  * Initial Release v1 modified Dualboot TWRP for the Oneplus 6/6t
  * Updated with Magisk 21.0

## Disclaimer
* I am not responsible for anything bad that happens to your device. Only experienced users should be using this mod
* This is no walk in the park mod. Although I have extensively tested it, there is always the possibility of a brick with anything that involves repartitioning. 
* Always make a Backup of all your Data before using this.
* **YOU'VE BEEN WARNED - Use at your own risk**

## Limitations / issues
* You can only set a PIN/Password in one of the ROMs not on both.

## Some other features/notes
* Can choose between stock layout, a/b userdata, a/b/c userdata where 'c' is a common data partition that'll show up in both roms, or StockSD layout which will create a seperate partion on Stock layout
* Option to choose between ext4 and f2fs
* Disables verity - fstabs are modified for dual boot and so this is a must unless you choose stock layout in which case it's optional
* Option to disable forced encryption
* Option to install magisk
* From Version 5 onwards there is a Dualboot Companion app added which can be used to reboot to other slots System or Recovery and it even will integrate in the reboot options from Android 11
* The Dualboot Companion app has also the ability to share apps Data between both ROMs (this is still in beta/test!) and to mount the other slots Data or System partition

## Common Data
* If you choose a/b/c layout - you'll have a/b userdata, but you'll also get a 3rd userdata partition called 'Common Data'
* The name 'Common Data' gives away its purpose - to store files that you'll access on both slots/roms. So stuff like zips, Pictures, Music, etc.
* In recovery, this shows up as another storage option for backup/restore and on your pc as well - your phone will have 'Common Storage' and 'Internal Storage'
  * The common data partition is mounted to /sdcard/CommonData
  * .nomedia file is placed in CommonData so files in it won't be picked up twice if you decide to mount over internal storage as outlined below
  * Furthermore, if your use case is like mine where my music files are in common data, you can make a 'mounts.txt' file in /datacommon containing a list of every FOLDER to mount directly over top of sdcard. So for example:<br/>
  /datacommon/Music -> /sdcard/Music
    * This of course mounts over anything there (overwrites it for as long as it's mounted) so make sure that you don't have the same folder in both datacommon and regular data
    * Note that there are 3 exceptions to this folder mounting rule:
      * All - if this is the FIRST line, ALL folders in datacommon will be mounted
      * Android
      * lost+found
    * The reasoning should be obvious - lost+found isn't something you should need to mess with and Android is for regular data partition only - that's OS specific and should be on separate slots
    * Note that you should have 1 folder listed on every line, for example:
      ```DCIM
      Music
      Pictures
      ViPER4AndroidFX
      ```
## Dualboot Companion app
* From Version 5 onwards there is a Dualboot Companion app (made by User Invernomut0) added to the Dualboot Recovery which can be used to reboot to other slots System or Recovery and it even will integrate in the reboot options from Android 11
* The Dualboot Companion app has also the ability to share apps Data between both ROMs (this is still in beta/test!) and to mount the other slots Data or System partition
  * It will create a folder called "DualBoot" inside the SDCARD with folders for "SDcard_a" / "SDcard_b" and "System_a" / "System_b"
  * If you are using the A/B/C Layout and have activated the "Share app" function inside the Companion app it will create a folder "SharedData" inside of your Commondata folder for the apps you selected to share the data throughout both ROMs with
  * The app also shows additional information about the installed ROM Version, Android Version, and security Patch for each ROM
  
## Flashing Instructions
* You MUST be booted into recovery already when flashing this Dualboot-mod.
* Since this modifies data - the zip CANNOT be on sdcard or data at all UNLESS you do not want to repartition/format
  * If you flash from data, the zip will copy itself to /tmp and instruct you to flash it from there
* Read through ALL the prompts - there's lots of options. 
* See more info here: https://github.com/Wishmasterflo/TWRP-DualBoot-Guac-Unified/blob/master/Text.md

## How to Flash Roms
* Nothing changes here except ONLY FLASH IN RECOVERY
  * Roms always flash to the opposite slot. Keep that in mind and you'll be fine
  * So don't take an OTA while booted - boot into recovery, switch slots, reboot into recovery, flash rom
* Normal flash procedure:
  * Boot into Recovery
  * Flash latest OOS (when doing a clean-flash!)
  * Flash rom (this will flash to the opposite slot)
  * Flash Dualboot-mod
  * Reboot to recovery
  * Flash everything else (GApps etc.)
  * Format data!

## Help! I Can't Boot!
* Usually this is because you switched roms without formatting data on that slot!


## Credits

* [Zackptg5](https://github.com/Zackptg5/TWRP-DualBoot-Guac-Unified)
* [Teamwin](https://github.com/TeamWin)
* [Mauronofrio](https://github.com/mauronofrio/android_device_oneplus_guacamole_unified_TWRP)
* [The OrangeFox Team]
* [TopJohnWu](https://github.com/topjohnwu/Magisk)
* [Ae3NerdGod](https://github.com/Ae3NerdGod)
* [Invernomut0](https://github.com/Invernomut0)
* [acuicultor](https://github.com/acuicultor)
* [Morphine1](https://github.com/morphine1)

## License

  MIT


Changelog.md

Changelog
---------
05/14/2021 - 3.5.2-9 v6
- Updated Dualboot Companion app to V2.9.1
DualBoot companion App V2.9.1 Changelog
- Optimized splash screen
- Fixed reboot time on Custom Rom in widget menu

04/09/2021 - 3.5.2-9 v5.3
- Updated TWRP to 3.5.2-9 (black Themed)
- Updated Dualboot Companion app to V2.8.9
  DualBoot companion App V2.8.9 Changelog
  - Added Splash screen at start
  - Optimized starting time
  - Added free space on CommonData
  - Donate button 
  - Update Russian translation
  - bug fix here and there
  - Fixed Shortcuts
  - Fixed colours in updated box
  - Updated Italian and Russian translation
  - new logo
  - new info section
  - fixed crash using shortcuts
  - shared app updated:
  - Data Restore feature (it will copy the shared data back to /data/data and dismount the bind mount)
  - Data Delete feature (it will delete the shared data and dismount the bind mount)
  - Info about shared data if they are in place in shared directory


03/18/2021 - 3.5.1-9 v5.2
- Updated with TWRP 3.5.1-9 (black Themed)


03/15/2021 - 3.5.0-9 v5.1
- Updated Dualboot Companion app to V2.8.5
- Added Dualboot with OrangeFox Recovery R11.0
- Added Dualboot Logo to OrangeFox Recovery
- You can now choose which recovery you want to use with the Dualboot. TWRP or OrangeFox Recovery
- Fixed bug in partition size details for 64GB Oneplus 6 devices
- Fixed error when flashing a ROM in OrangeFox recovery
- Fixed hanging/timeout when rebooting to System or Recovery on Android 11 (works only when using reboot function from inside the Dualboot Companion app)


03/05/2021 - 3.5.0-9 v5
- Added Dualboot Companion app created by User Invernomut0
- small bugfixes and corrections


02/07/2021 - 3.5.0-9 v4.4
- Fixed bug which caused that the quickmode options were not working


01/17/2021 - 3.5.0-9 v4.3
- Updated with Magisk 21.4


01/16/2021 - 3.5.0-9 v4.2
- Updated with Magisk 21.3


01/08/2021 - 3.5.0-9 v4.1
- Cleaned up stuff which is not needed anymore
- Fixed bugs that were existing in the repartitioning part
- Updated with official TWRP 3.5.0-9 and added black Theme to it
- Added option to create a seperate partition called StockSD when on Stock Layout


12/29/2020 - 3.4.0-0 v4
- Updated with Magisk 21.2
- Removed Magisk Canary 20422 since Magisk 21.2 is now working with TWRP and Android 11

12/01/2020 - 3.4.0-0 v3.3
- Added to flash Magisk Canary 20422 on Android 11 automatically
- Added mount file to mount Android Data to Commondata (still Beta/Test)


11/14/2020 - 3.4.0-0 v3.2
- Updated with Magisk 21.1
- Prepared script for some future features


10/30/2020 - 3.4.0-0 v3.1
- Removed unnessessary stuff which was left from the Oneplus 7 devices
- Fixed Common data mount script for Android 11
- Magisk 20.4
- Dualboot-TWRP can be flashed from USB Storage as well now
 

10/20/2020 - 3.4.0-0 v3
- Added Blackout TWRP Themed by acuicultor and Morphine1
- Added support/detection for 64GB Oneplus 6 enchilada devices
- Added that Magisk 21.0 does not get flashed on Android 11 ROMs (as that is not working now!)
- removed automatic launch of Common data mount script on Android 11 as that caused strange behaviour when rebooting the phone!
- Some more small adjustments in the flashing script
- Adjusted sizes for the userdata partitions when using A/B/C Layout as follows: 
  64GB    16, 24 GB partition size for A/B/C Layout
  128GB   16, 32, 40, 45 GB partition size for A/B/C Layout
  256GB   16, 32, 64, 96 GB partition size for A/B/C Layout


10/17/2020 - 3.4.0-0 v2.3
- Fixed mounting of Common data partition


10/15/2020 - 3.4.0-0 v2.2
- Fixed better detection of Android 11 Installed ROMS
- Build Dualboot Installer for Oneplus 6 enchilada


10/13/2020 - 3.4.0-0 v2.1
- Fixed hanging when on Android 11


10/12/2020 - 3.4.0-0 v2
- Fixed automatic Formatting and creating of the partitions
- added quickmode options which can be provided directly to the ZIP file name
- A lot of other bugfixes and improvments


10/06/2020 - 3.4.0-0 v1
- Initial Release v1 modified Dualboot TWRP for the Oneplus 6/6t
- Updated with Magisk 21.0




tools/functions.sh
#!/sbin/sh

chooseport() {
  while true; do
    getevent -lc 1 2>&1 | grep VOLUME | grep " DOWN" > $tmp/events
    if (`cat $tmp/events 2>/dev/null | grep VOLUME >/dev/null`); then
      break
    fi
  done
  if (`cat $tmp/events 2>/dev/null | grep VOLUMEUP >/dev/null`); then
    return 0
  else
    return 1
  fi
}

unmountAllAndRefreshPartitions() {
	mount | grep /dev/block/ | while read -r line ; do
		thispart=`echo "$line" | awk '{ print $3 }'`
		umount -lf $thispart
		sleep 0.5
	done
	stop sbinqseecomd
	sleep 1
	kill `pidof qseecomd` 2>/dev/null
	blockdev --rereadpt /dev/block/sda
  sleep 0.5
}

select_size() {
  ui_print " "
  ui_print "- Supported Partition Sizes for userdata:"
  [ $totsize -gt 128 ] && local sizes="16, 32, 64, or 96 GB" || local sizes="16, 32, 40, or 45 GB"
  ui_print "  $sizes"
  ui_print "  Common Data takes up whatever is left"
  sleep 1
  ui_print " "
  ui_print "  Choose Userdata Size:"
  for i in 16 32 40 45 64 96; do
    [ "$sizes" == "or $i GB" ] && { size=$i; return 0; }
    sizes="$(echo $sizes | sed -r "s/^$i, |^$i //g")"
    ui_print "  Vol+ $i GB; Vol- $sizes"
    chooseport && { size=$i; return 0; }
    ui_print " "
  done
}

select_format() {
  ui_print " "
  ui_print "- Ext4 or F2FS?"
  ui_print "  Note that your kernel must be f2fs compatible"
  ui_print "  Vol+ Ext4 (Recommended), Vol- F2FS"
  if [ "$layout" == "stock" ]; then
    chooseport && typea="ext4" || typea="f2fs"
  elif [ "$layout" == "stocksd" ]; then
    chooseport && typea="ext4" || typea="f2fs"
  else
    ui_print " "
    [ "$layout" == "a/b/c" ] && local slots="a b c" || local slots="a b"
    for i in $slots; do
      ui_print "  Slot $i?"
      chooseport && eval type$i="ext4" || eval type$i="f2fs"
    done
  fi
}

change_part() {
  local action=$1 partnum=$2 parttype=$3
  [ -z "$partnum" ] && abort "Partitioning error!"
  case $action in
    "delete") sgdisk /dev/block/sda --delete $partnum;;
    "new") sgdisk /dev/block/sda --new=$partnum;;
    "change-name") sgdisk /dev/block/sda --change-name=$partnum;;
    "format")
      local partsize=$(sgdisk --print /dev/block/sda | grep "^ *$partnum" | awk '{print $3-$2+1}')
      [ -z "$partsize" ] && abort "Formatting error!"
      local partname="$(sgdisk --print /dev/block/sda | grep "^ *$partnum" | awk '{gsub("_.*","",$7); print $7}' 2>/dev/null)"
      [ -z "$partname" ] && abort "Formatting error!"
      if [ "$parttype" == "f2fs" ]; then
        mkfs.f2fs -d1 -f -O encrypt -O quota -O verity -w 4096 /dev/block/sda$partnum $partsize
        sload.f2fs -t /data /dev/block/sda$partnum
      else
        mke2fs -F -F -t ext4 -b 4096 /dev/block/sda$partnum $partsize
        e2fsdroid -e -S /file_contexts -a /data /dev/block/sda$partnum
      fi
      ;;
	*) abort "Partitioning error!";;
  esac
  [ $? -eq 0 ] || abort "Partitioning error! You'll need to fix it manually!"
}

stock_userdata() {
  if [ "$curlayout" == "a/b" ]; then
      userdata_a_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata_a')
      userdata_a_partnum=$(echo "$userdata_a_partline" | awk '{ print $1 }')
      userdata_a_partend=$(echo "$userdata_a_partline" | awk '{ print $3 }')
      userdata_b_partline=$(sgdisk --print /dev/block/sda | grep -i userdata_b)
      userdata_b_partnum=$(echo "$userdata_b_partline" | awk '{ print $1 }')
	  userdata_b_partend=$(echo "$userdata_b_partline" | awk '{ print $3 }')
	  change_part delete $userdata_a_partnum
	  change_part delete $userdata_b_partnum
	  userdata_partend=$userdata_b_partend
      change_part new $userdata_partnum:$userdata_partstart:$userdata_partend
  fi	  
  if [ "$curlayout" == "stocksd" ]; then
      userdata_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata')
      userdata_partnum=$(echo "$userdata_partline" | awk '{ print $1 }')
      userdata_partend=$(echo "$userdata_partline" | awk '{ print $3 }')
      userdata_stocksd_partline=$(sgdisk --print /dev/block/sda | grep -i 'stocksd')
      userdata_stocksd_partnum=$(echo "$userdata_stocksd_partline" | awk '{ print $1 }')
      userdata_stocksd_partend=$(echo "$userdata_stocksd_partline" | awk '{ print $3 }')
      change_part delete $userdata_stocksd_partnum
      change_part delete $userdata_partnum
	  userdata_partend=$userdata_stocksd_partend
      change_part new $userdata_partnum:$userdata_partstart:$userdata_partend    	  
   fi  
   if [ "$curlayout" == "a/b/c" ]; then
      userdata_a_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata_a')
      userdata_a_partnum=$(echo "$userdata_a_partline" | awk '{ print $1 }')
      userdata_a_partend=$(echo "$userdata_a_partline" | awk '{ print $3 }')
      userdata_b_partline=$(sgdisk --print /dev/block/sda | grep -i userdata_b)
      userdata_b_partnum=$(echo "$userdata_b_partline" | awk '{ print $1 }')
	  userdata_b_partend=$(echo "$userdata_b_partline" | awk '{ print $3 }')
	  userdata_c_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata2')
      userdata_c_partnum=$(echo "$userdata_c_partline" | awk '{ print $1 }')
      userdata_c_partend=$(echo "$userdata_c_partline" | awk '{ print $3 }')
      change_part delete $userdata_a_partnum
	  change_part delete $userdata_b_partnum   
      userdata_partend=$userdata_c_partend
	  change_part delete $userdata_c_partnum
	  change_part new $userdata_partnum:$userdata_partstart:$userdata_partend
    fi
   change_part change-name $userdata_partnum:userdata
}

slot_userdata() {
	if [ "$layout" == "stocksd" ]; then
    typea="ext4"
	typed="ext4"
    userdata_length=`echo $((userdata_partend-userdata_partstart))`
    userdata_new=`echo $((userdata_length / 3))`
    userdata_new_partend=`echo $((userdata_partstart+userdata_new))`
    userdata_stocksd_partstart=`echo $((userdata_new_partend+1))`
    userdata_stocksd_partnum=`echo $((userdata_partnum+1))`
    change_part delete $userdata_partnum
    change_part new $userdata_partnum:$userdata_partstart:$userdata_new_partend
    change_part change-name $userdata_partnum:userdata
    change_part new $userdata_stocksd_partnum:$userdata_stocksd_partstart:$userdata_partend
    change_part change-name $userdata_stocksd_partnum:stocksd
	
	elif [ "$layout" == "a/b/c" ]; then
		# Calculate size for $size partition with 4k logical sector size
    userdata_newpartsize=$((size * 1024 * 1024 / 4))
    userdata_a_partend=`echo $((userdata_partstart+userdata_newpartsize))`

    userdata_b_partstart=`echo $((userdata_a_partend+1))`
    userdata_b_partend=`echo $((userdata_b_partstart+userdata_newpartsize))`
    userdata_b_partnum=$(echo $((`sgdisk --print /dev/block/sda | awk 'END{print $1}'`+1)))
	
    userdata_c_partstart=`echo $((userdata_b_partend+1))`
    userdata_c_partnum=`echo $((userdata_b_partnum+1))`
	
	change_part delete $userdata_partnum
  	change_part new $userdata_partnum:$userdata_partstart:$userdata_a_partend 
    change_part change-name $userdata_partnum:userdata_a
	change_part new $userdata_b_partnum:$userdata_b_partstart:$userdata_b_partend
    change_part change-name $userdata_b_partnum:userdata_b
  
	else
		# Cut it in half
	userdata_length=`echo $((userdata_partend-userdata_partstart))`
    userdata_newpartsize=`echo $((userdata_length / 2))`
	userdata_a_partend=`echo $((userdata_partstart+userdata_newpartsize))`

    userdata_b_partstart=`echo $((userdata_a_partend+1))`
    userdata_b_partend=$userdata_partend
    userdata_b_partnum=`echo $((userdata_partnum+1))`
	
	change_part delete $userdata_partnum
  	change_part new $userdata_partnum:$userdata_partstart:$userdata_a_partend 
    change_part change-name $userdata_partnum:userdata_a
	change_part new $userdata_b_partnum:$userdata_b_partstart:$userdata_b_partend
    change_part change-name $userdata_b_partnum:userdata_b
    fi

	if [ "$layout" == "a/b/c" ]; then
    change_part new $userdata_c_partnum:$userdata_c_partstart:$userdata_partend
    change_part change-name $userdata_c_partnum:userdata2
  fi
  blockdev --rereadpt /dev/block/sda
	sleep 0.5
	ui_print "  Formatting new userdata partitions"
	change_part format $userdata_partnum $typea
  [ "$layout" == "stocksd" ] && change_part format $userdata_stocksd_partnum $typed
  [ "$layout" == "a/b" ] && change_part format $userdata_b_partnum $typeb
  [ "$layout" == "a/b/c" ] && change_part format $userdata_b_partnum $typeb
  [ "$layout" == "a/b/c" ] && change_part format $userdata_c_partnum $typec
  ui_print "  Userdata has been partitioned to $layout"
}

repartition_userdata() {
  ui_print " "
   if [ "$curlayout" == "stocksd" ]; then
        userdata_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata')	
   elif [ "$curlayout" == "stock" ]; then
		userdata_partline=$(sgdisk --print /dev/block/sda | grep -i 'userdata$')
  else
		userdata_partline=$(sgdisk --print /dev/block/sda | grep -i userdata_a)
	fi
	userdata_partnum=$(echo "$userdata_partline" | awk '{ print $1 }')
	userdata_partstart=$(echo "$userdata_partline" | awk '{ print $2 }')
	userdata_partend=$(echo "$userdata_partline" | awk '{ print $3 }')
	userdata_partname=$(echo "$userdata_partline" | awk '{ print $7 }')
  case $layout in
    "stock") 
      ui_print "- Repartitioning userdata back to stock"
      stock_userdata   
      blockdev --rereadpt /dev/block/sda
      sleep 0.5
      ui_print "  Formatting userdata"
      change_part format $userdata_partnum $typea
      ui_print "  Userdata has been repartitioned back to stock"
      ;;
    *)
      ui_print "- Repartitioning userdata to $layout"
      [ "$curlayout" == "stock" ] || stock_userdata
      slot_userdata
      ;;
  esac
}

reset_slot() {
  slot=$(getprop ro.boot.slot_suffix 2>/dev/null)
  test "$slot" || slot=$(grep -o 'androidboot.slot_suffix=.*$' /proc/cmdline | cut -d\  -f1 | cut -d= -f2)
  if [ ! "$slot" ]; then
    slot=$(getprop ro.boot.slot 2>/dev/null)
    test "$slot" || slot=$(grep -o 'androidboot.slot=.*$' /proc/cmdline | cut -d\  -f1 | cut -d= -f2)
    test "$slot" && slot=_$slot
  fi;
  if [ "$slot" ]; then
    case $slot_select in
      inactive)
        case $slot in
          _a) slot=_b;;
          _b) slot=_a;;
        esac
      ;;
    esac
  fi
  [ -z $slot ] && abort "Unable to determine active boot slot! Aborting!"
}

patch_fstabs() {
	ui_print "  Patching fstabs:"
	for i in $FSTABS; do
		[ -f $i ] || continue
		ui_print "    $i"
    local perm="$(ls -Z $i | awk '{print $1}')"
		sed -ri "
      s/,verifyatboot|verifyatboot,|verifyatboot\b//g
      s/,verify|verify,|verify\b//g
      s/,avb_keys|avb_keys,|avb_keys\b//g
      s/,avb|avb,|avb\b//g
      s/,support_scfs|support_scfs,|support_scfs\b//g
      s/,fsverity|fsverity,|fsverity\b//g
    " "$i"
    $KEEPFORCEENCRYPT || sed -i "s/fileencryption=/=/g" $i
    sed -i "/data2/d" $i
		sed -i "/name\/userdata / s/wait,slotselect/wait/" $i
    while true; do
      [ "$(tail -n1 $i)" ] && { echo >> $i; break; } || sed -i '$d' $i
    done    
	[ "$layout" == "stock" ] || sed -i "/name\/userdata / s/wait,/wait,slotselect,/" $i
		chcon $perm $i
	done
}

#########################################
#
# Magisk General Utility Functions
# by topjohnwu
#
#########################################

binary_patches() { # Modified
	for dt in dtb kernel_dtb extra recovery_dtbo; do
		[ -f $dt ] && magiskboot dtb $dt patch && ui_print "    Patch fstab in $dt"
	done

	if [ -f kernel ]; then
    if $magisk; then
      # Force kernel to load rootfs
      # skip_initramfs -> want_initramfs
      magiskboot hexpatch kernel \
      736B69705F696E697472616D667300 \
      77616E745F696E697472616D667300
    else
      # want_initramfs -> skip_initramfs
      magiskboot hexpatch kernel \
      77616E745F696E697472616D6673 \
      736B69705F696E697472616D6673
    fi
	fi
}

###################
# Helper Functions
###################

toupper() {
  echo "$@" | tr '[:lower:]' '[:upper:]'
}

grep_prop() {
  local REGEX="s/^$1=//p"
  shift
  local FILES=$@
  [ -z "$FILES" ] && FILES='/system/build.prop'
  sed -n "$REGEX" $FILES 2>/dev/null | head -n 1
}

#######################
# Installation Related
#######################

find_block() {
  for BLOCK in "$@"; do
    DEVICE=`find /dev/block -type l -iname $BLOCK | head -n 1` 2>/dev/null
    if [ ! -z $DEVICE ]; then
      readlink -f $DEVICE
      return 0
    fi
  done
  # Fallback by parsing sysfs uevents
  for uevent in /sys/dev/block/*/uevent; do
    local DEVNAME=`grep_prop DEVNAME $uevent`
    local PARTNAME=`grep_prop PARTNAME $uevent`
    for BLOCK in "$@"; do
      if [ "`toupper $BLOCK`" = "`toupper $PARTNAME`" ]; then
        echo /dev/block/$DEVNAME
        return 0
      fi
    done
  done
  return 1
}

patch_dtb_partitions() { # Modified
  set +x
  local result=1
  for name in dtb dtbo; do
    local IMAGE=`find_block $name$slot`
    if [ ! -z $IMAGE ]; then
      ui_print "    $name image: $IMAGE"
      if magiskboot dtb $IMAGE patch dt.patched; then
        result=0
				$magisk && { ui_print "    Backing up stock $name image"
        cat $IMAGE > stock_${name}.img; }
        ui_print "    Flashing patched $name"
        cat dt.patched /dev/zero > $IMAGE
        rm -f dt.patched
      fi
    fi
  done
  set -x
  return $result
}

run_migrations() {
  local LOCSHA1
  local TARGET
  # Legacy app installation
  local BACKUP=/data/adb/magisk/stock_boot*.gz
  if [ -f $BACKUP ]; then
    cp $BACKUP /data
    rm -f $BACKUP
  fi

  # Legacy backup
  for gz in /data/stock_boot*.gz; do
    [ -f $gz ] || break
    LOCSHA1=`basename $gz | sed -e 's/stock_boot_//' -e 's/.img.gz//'`
    [ -z $LOCSHA1 ] && break
    mkdir /data/magisk_backup_${LOCSHA1} 2>/dev/null
    mv $gz /data/magisk_backup_${LOCSHA1}/boot.img.gz
  done

  # Stock backups
  LOCSHA1=$SHA1
  for name in boot dtb dtbo; do
    BACKUP=/data/adb/magisk/stock_${name}.img
    [ -f $BACKUP ] || continue
    if [ $name = 'boot' ]; then
      LOCSHA1=`$MAGISKBIN/magiskboot sha1 $BACKUP`
      mkdir /data/magisk_backup_${LOCSHA1} 2>/dev/null
    fi
    TARGET=/data/magisk_backup_${LOCSHA1}/${name}.img
    cp $BACKUP $TARGET
    rm -f $BACKUP
    gzip -9f $TARGET
  done
}

#################
# Module Related
#################

set_perm() {
  chown $2:$3 $1 || return 1
  chmod $4 $1 || return 1
  CON=$5
  [ -z $CON ] && CON=u:object_r:system_file:s0
  chcon $CON $1 || return 1
}
